### ISSUE TITLE
Support for EasyAuth / Authenticaton Block in azurerm_container_app resource

### ISSUE THREAD

**Issue by Lachlan-White** [hashicorp/terraform-provider-azurerm#22213]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

As Azure Container Apps become more frequently deployed the capability of Easy Auth or Authentication settings as a configurable block similar to that `azurerm_linux_web_app` is highly desirable. 

Enabling the end to end container app to be managed by the AzureRM provider instead of cobbling together AzAPI or AzCLI workflows in addition to the resource definiton.

### New or Affected Resource(s)/Data Source(s)

azurerm_container_app

### Potential Terraform Configuration

```hcl
resource "azurerm_container_app" "example" {
  name                         = "example-app"
  container_app_environment_id = azurerm_container_app_environment.example.id
  resource_group_name          = azurerm_resource_group.example.name
  revision_mode                = "Single"

  template {
    container {
      name   = "examplecontainerapp"
      image  = "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest"
      cpu    = 0.25
      memory = "0.5Gi"
    }
    auth_settings {
      enabled = true
      require_authentication = true
      unauthenticated_action = "Return403"
      default_provider = "azureactivedirectory"
        microsoft {
            client_id = "xxx-xxx-xxx-xxx"
            client_secret_name_setting = "microsoft-identity-provider-secret"
            supported_account_types = "currenttenant"
        }
    }
  }
}
```

### References

Microsoft Documentation (Learn) Reference:
https://learn.microsoft.com/en-us/azure/container-apps/authentication-azure-active-directory

Linux Web App Auth Settings for Reference:
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_web_app#auth_settings_v2

**Comment by fredlemieux** [comment]
This would be a great feature, any chance of moving this up the priority list?

**Comment by DeDeckerThomas** [comment]
This would be a very great feature to have indeed!

**Comment by cfarrugia** [comment]
Bumped into this issue and pretty disappointed as I have all my infra nicely terraformed. However I managed to go around it using a provisioner within the azurerm_container_app definition like this:

provisioner "local-exec" {
    command = "az containerapp auth update --name aca-feed-export-api-${var.name_of_app_resrouce} --resource-group ${var.rgname} --unauthenticated-client-action AllowAnonymous"

In my case i needed nothing but allowing anon access, and this worked brilliantly.

**Comment by dhduvall** [comment]
I put together a solution using the `azapi` provider, if it's useful to anyone else.

```terraform
# https://learn.microsoft.com/en-us/rest/api/containerapps/container-apps-auth-configs/create-or-update
resource "azapi_resource_action" "my_app_auth" {
  type        = "Microsoft.App/containerApps/authConfigs@2024-03-01"
  resource_id = "${azurerm_container_app.my_app.id}/authConfigs/current"
  method      = "PUT"

  body = jsonencode({
    location = azurerm_resource_group.ev_rg.location
    properties = {
      globalValidation = {
        redirectToProvider          = "azureactivedirectory"
        unauthenticatedClientAction = "RedirectToLoginPage"
      }
      identityProviders = {
        azureActiveDirectory = {
          registration = {
            clientId                = azuread_application_registration.my_app.client_id
            clientSecretSettingName = "microsoft-provider-authentication-secret"
            openIdIssuer            = "https://sts.windows.net/${data.azurerm_subscription.current.tenant_id}/v2.0"
          }
          validation = {
            defaultAuthorizationPolicy = {
              allowedApplications = [
                azuread_application_registration.my_app.client_id,
              ]
            }
          }
        }
      }
      platform = {
        enabled = true
      }
    }
  })
```

**Comment by EpicWink** [comment]
Note you may want: `depends_on = [azurerm_container_app.my_app]` if Terraform is having a hard time figuring out dependencies

---


<details><summary>Resource for APIs (click to expand)</summary>

```terraform
# See: https://learn.microsoft.com/en-us/rest/api/resource-manager/containerapps/container-apps-auth-configs/create-or-update
resource "azapi_resource_action" "my_app_auth" {
  depends_on = [azurerm_container_app.my_app]

  type        = "Microsoft.App/containerApps/authConfigs@2024-03-01"
  resource_id = "${azurerm_container_app.my_app.id}/authConfigs/current"
  method      = "PUT"

  body = { # wrap in jsondecode if using 'azapi' v1
    location = azurerm_container_app.my_app.location
    properties = {
      globalValidation = {
        unauthenticatedClientAction = "Return401"
      }
      identityProviders = {
        azureActiveDirectory = {
          enabled = true
          registration = {
            clientId                = azuread_application.my_app.client_id
            clientSecretSettingName = "microsoft-provider-authentication-secret"
            openIdIssuer            = "https://sts.windows.net/${data.azuread_client_config.current.tenant_id}/v2.0"
          }
          validation = {
            allowedAudiences = [tolist(azuread_application.app.identifier_uris)[0]]
            defaultAuthorizationPolicy = {
              allowedApplications = [
                azuread_application.my_app.client_id,
              ]
            }
          }
        }
      }
      platform = {
        enabled = true
      }
    }
  }
}
```

</details>

**Comment by jason-johnson** [comment]
> Note you likely want: `depends_on = [azurerm_container_app.my_app]`
> 
> Resource for APIs (click to expand)

No, the resource already references `azurerm_container_app.my_app` in the `resource_id` field.

**Comment by cssoftware-main** [comment]
Any updates on this?

**Comment by tombrewer-onwave** [comment]
it's been a minute - any kind of ETA?

**Comment by Techital-jonathan** [comment]
This is sorely needed. I don't like custom "scripts" nor manual click ops :).

**Comment by jr-thirdfin** [comment]
It's better to use something like `azapi_resource` (Declarative) rather than `azapi_resource_action` (Imperative) as you get to manage the resource. The below just disables but is a better pattern than mixing imperative in your TF.

```
resource "azapi_resource" "app_auth" {
  type      = "Microsoft.App/containerApps/authConfigs@2025-01-01"
  name      = "current"
  parent_id = azurerm_container_app.app.id

  body = {
    properties = {
      platform = {
        enabled = false
      },
      globalValidation = { unauthenticatedClientAction = "AllowAnonymous" }
    }
  }
}
```

Plus 1 on adding this

**Comment by Jay-diehl** [comment]
Would be a helpful addition +1

**Comment by victorfrye** [comment]
Definitely would be nice addition to azurerm provider +1

**Comment by jason-johnson** [comment]
Any update?  The other platforms like Azure Functions, Azure Web Apps, etc. have this and the API clearly supports it.  Is there something making it harder to support with ACA?

**Comment by ArcTheMaster** [comment]
Hi there,
Also joining the party here, the feature can be a very good addition to azurerm for ACA. 
What are the current blockers for implementing this kind of new resource ?

**Comment by ffloimair** [comment]
Same here, desperately waiting for it.

**Comment by egorshulga** [comment]
The EasyAuth is treated as a separate resource in Azure, thus, it is easy to create it with a AzAPI (how was also suggested above). Here is a complete working snippet:

```terraform
data "azuread_application" "connectionToIdentityProvider" {
  object_id = YOUR_APPLICATION_OBJECT_ID
}

resource "azurerm_user_assigned_identity" "authIdentity" {
  resource_group_name = RESOURCE_GROUP_NAME
  location            = RESOURCE_GROUP_LOCATION
  name                = "container-app-auth"
}

# The Service Principal creating this resource must be the Owner
# of the Application Registration so that it could create the federated credential.
resource "azuread_application_federated_identity_credential" "federatedCredential" {
  display_name   = "container-app-auth"
  description    = "Used for configuring built-in authentication in Container App"
  application_id = data.azuread_application.connectionToIdentityProvider.id
  issuer         = "https://login.microsoftonline.com/${data.azurerm_client_config.current.tenant_id}/v2.0"
  subject        = azurerm_user_assigned_identity.authIdentity.principal_id
  audiences      = ["api://AzureADTokenExchange"]
}

resource "azapi_resource" "containerAppAuthentication" {
  type      = "Microsoft.App/containerApps/authConfigs@2025-02-02-preview"
  name      = "current"
  parent_id = azurerm_container_app.containerApp.id # ID OF THE CONTAINER APP GOES HERE

  body = {
    properties = {
      platform = { enabled = true }
      globalValidation = {
        unauthenticatedClientAction = "Return401"
        excludedPaths               = ["/ping"]
      }

      identityProviders = {
        azureActiveDirectory = {
          enabled = true
          registration = {
            openIdIssuer            = "https://sts.windows.net/${data.azurerm_client_config.current.tenant_id}/v2.0"
            clientId                = data.azuread_application.connectionToIdentityProvider.client_id
            clientSecretSettingName = "override-use-mi-fic-assertion-client-id"
          }
        }
      }
    }
  }
}

```

upd: Last step: don't forget to add a secret / app setting with the name `override-use-mi-fic-assertion-client-id` and the value `azurerm_user_assigned_identity.authIdentity.client_id`

**Comment by kkarballof** [comment]
> The EasyAuth is treated as a separate resource in Azure, thus, it is easy to create it with a AzAPI (how was also suggested above). Here is a complete working snippet:
> 
> data "azuread_application" "connectionToIdentityProvider" {
>   object_id = YOUR_APPLICATION_OBJECT_ID
> }
> 
> resource "azurerm_user_assigned_identity" "authIdentity" {
>   resource_group_name = RESOURCE_GROUP_NAME
>   location            = RESOURCE_GROUP_LOCATION
>   name                = "container-app-auth"
> }
> 
> # The Service Principal creating this resource must be the Owner
> # of the Application Registration so that it could create the federated credential.
> resource "azuread_application_federated_identity_credential" "federatedCredential" {
>   display_name   = "container-app-auth"
>   description    = "Used for configuring built-in authentication in Container App"
>   application_id = data.azuread_application.connectionToIdentityProvider.id
>   issuer         = "https://login.microsoftonline.com/${data.azurerm_client_config.current.tenant_id}/v2.0"
>   subject        = azurerm_user_assigned_identity.authIdentity.principal_id
>   audiences      = ["api://AzureADTokenExchange"]
> }
> 
> resource "azapi_resource" "containerAppAuthentication" {
>   type      = "Microsoft.App/containerApps/authConfigs@2025-02-02-preview"
>   name      = "current"
>   parent_id = azurerm_container_app.containerApp.id # ID OF THE CONTAINER APP GOES HERE
> 
>   body = {
>     properties = {
>       platform = { enabled = true }
>       globalValidation = {
>         unauthenticatedClientAction = "Return401"
>         excludedPaths               = ["/ping"]
>       }
> 
>       identityProviders = {
>         azureActiveDirectory = {
>           enabled = true
>           registration = {
>             openIdIssuer            = "https://sts.windows.net/${data.azurerm_client_config.current.tenant_id}/v2.0"
>             clientId                = data.azuread_application.connectionToIdentityProvider.client_id
>             clientSecretSettingName = "override-use-mi-fic-assertion-client-id"
>           }
>         }
>       }
>     }
>   }
> }

@egorshulga 
For some reason I am being forced to create an app registration secret anyway. I am missing container apps env variables? Here is the error the authentication blade is throwing "Your app references secrets not found on your Entra ID application. Authentication on your app may be in a broken state."

**Comment by kkarballof** [comment]
For some reason I am being forced to create an app registration secret anyway. I am missing container apps env variables? Here is the error the authentication blade is throwing "Your app references secrets not found on your Entra ID application. Authentication on your app may be in a broken state."

**Comment by egorshulga** [comment]
@kkarballof indeed, there additionally should be a Container App's Secret named `override-use-mi-fic-assertion-client-id` with the value of Client ID of the created managed identity ([here](https://learn.microsoft.com/en-us/azure/app-service/configure-authentication-provider-aad?tabs=workforce-configuration#use-a-managed-identity-instead-of-a-secret-preview) are the docs).

Btw, this works exactly the same for App Services - the same app setting `override-use-mi-fic-assertion-client-id` would enable the same auth.

**Comment by kkarballof** [comment]
@egorshulga That's what I did but but I only get 401 and this pops over my auth blade => "Your app references secrets not found on your Entra ID application. Authentication on your app may be in a broken state."

**Comment by alexb-iolo** [comment]
Still an issue!

**Comment by jason-johnson** [comment]
> The EasyAuth is treated as a separate resource in Azure, thus, it is easy to create it with a AzAPI (how was also suggested above). Here is a complete working snippet:
> 
> data "azuread_application" "connectionToIdentityProvider" {
>   object_id = YOUR_APPLICATION_OBJECT_ID
> }
> 
> resource "azurerm_user_assigned_identity" "authIdentity" {
>   resource_group_name = RESOURCE_GROUP_NAME
>   location            = RESOURCE_GROUP_LOCATION
>   name                = "container-app-auth"
> }
> 
> # The Service Principal creating this resource must be the Owner
> # of the Application Registration so that it could create the federated credential.
> resource "azuread_application_federated_identity_credential" "federatedCredential" {
>   display_name   = "container-app-auth"
>   description    = "Used for configuring built-in authentication in Container App"
>   application_id = data.azuread_application.connectionToIdentityProvider.id
>   issuer         = "https://login.microsoftonline.com/${data.azurerm_client_config.current.tenant_id}/v2.0"
>   subject        = azurerm_user_assigned_identity.authIdentity.principal_id
>   audiences      = ["api://AzureADTokenExchange"]
> }
> 
> resource "azapi_resource" "containerAppAuthentication" {
>   type      = "Microsoft.App/containerApps/authConfigs@2025-02-02-preview"
>   name      = "current"
>   parent_id = azurerm_container_app.containerApp.id # ID OF THE CONTAINER APP GOES HERE
> 
>   body = {
>     properties = {
>       platform = { enabled = true }
>       globalValidation = {
>         unauthenticatedClientAction = "Return401"
>         excludedPaths               = ["/ping"]
>       }
> 
>       identityProviders = {
>         azureActiveDirectory = {
>           enabled = true
>           registration = {
>             openIdIssuer            = "https://sts.windows.net/${data.azurerm_client_config.current.tenant_id}/v2.0"
>             clientId                = data.azuread_application.connectionToIdentityProvider.client_id
>             clientSecretSettingName = "override-use-mi-fic-assertion-client-id"
>           }
>         }
>       }
>     }
>   }
> }
> upd: Last step: don't forget to add a secret / app setting with the name `override-use-mi-fic-assertion-client-id` and the value `azurerm_user_assigned_identity.authIdentity.client_id`

Does this mean this will never be supported by the ACA resources the way it is in azure web app and azure function resources?  Azapi can do anything that can be done in ARM but generally the azurerm resources are nicer to use.

**Comment by Tolbin400** [comment]
I think asking people to use the azapi to patch the resource adds complexity for customers when consuming the module using the azurerm provider.

Would it make sense for it to be a separate resource though for container Apps in the azurerm provider as microsoft treat it as a separate resource

**Comment by egorshulga** [comment]
Of course that would be easier to have it in azurerm 😁 The question is if you want to have it now, or implement the feature yourself, or wait for eternity
